### PR TITLE
Add to_dict to ParameterConfig

### LINKED ISSUE #5322: Add to/from-json for Field
Currently the field information is saved explicitly in: https://github.com/equinor/ert/blob/daea5dfe98e874b157491162eccf93115d1189a2/src/ert/storage/local_experiment.py#L143 which is called in the constructor: https://github.com/equinor/ert/blob/daea5dfe98e874b157491162eccf93115d1189a2/src/ert/storage/local_experiment.py#L108-L109, we should instead save this on the experiment in a `parameters.json` file with a structure something like:
```json
{
    "parameter_1": {
        "name": "parameter_1",
        "type": "Field",
        "nx": 12
        ...
    },
    "parameter_2": {
        "name": "parameter_2",
        "type": "Surface",
        "rotation": 1
        ...
    }
}

```
where `Field` has a `to_json` and `from_json` we can use to save/load metadata.

### PR DESCRIPTION
**Pull request by frode-aarstad**
**Issue**
Resolves #5322 

**Approach**
_Short description of the approach_

## Pre review checklist

- [ ] Added appropriate release note label
- [ ] PR title captures the intent of the changes, and is fitting for release notes.
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [ ] Updated documentation
- [ ] Ensured new behaviour is tested

Adding labels helps the maintainers when writing release notes. This is the [list of release note labels](https://github.com/equinor/ert/labels?q=release-notes).

### CODE REVIEW

File: src/ert/_c_wrappers/enkf/config/parameter_config.py [hunk 1]
```diff
@@ -1,16 +1,29 @@
 from __future__ import annotations
 
+import dataclasses
```
Comment (dafeda) [inline src/ert/_c_wrappers/enkf/config/parameter_config.py:3]: A general comment to `parameter_config.py` is that it is not a cwrapper and should be moved somewhere else.
No need to do that in this PR, so I have created an issue for it: https://github.com/equinor/ert/issues/5395
I also think that we should add type-hints even though cwrappers are not checked by mypy.

File: src/ert/_c_wrappers/enkf/config/parameter_config.py [hunk 2]
```diff
@@ -1,16 +1,29 @@
...
 
 
-@dataclass
+class CustomDict(dict):
+    """Used for converting types that can not be serialized
+    directly to json
+    """
+
+    def __init__(self, data):
+        for i, e in enumerate(data):
```
Comment (dafeda) [inline src/ert/_c_wrappers/enkf/config/parameter_config.py:20]: I prefer `for i, (key, value) in enumerate(data)` but not that important.

File: src/ert/_c_wrappers/enkf/config/parameter_config.py [hunk 3]
```diff
@@ -1,16 +1,29 @@
...
 
 if TYPE_CHECKING:
-    from pathlib import Path
+    from typing import Any
 
     from ert.storage import EnsembleAccessor, EnsembleReader
 
 
-@dataclass
+class CustomDict(dict):
```
Comment (dafeda) [inline src/ert/_c_wrappers/enkf/config/parameter_config.py:14]: Perhaps a more explicit name is `PathStrDict`.

File: src/ert/storage/local_experiment.py [hunk 4]
```diff
@@ -59,11 +60,11 @@ def gen_kw_info(self) -> Dict[str, Any]:
...
 
     @property
-    def field_info(self) -> Dict[str, Any]:
+    def parameter_info(self) -> Dict[str, Any]:
         info: Dict[str, Any]
-        path = self.mount_point / "field-info.json"
+        path = self.mount_point / self._parameter_file
         if not path.exists():
-            raise ValueError("No field info exists")
+            raise ValueError("No parameter.json exists")
```
Comment (dafeda) [inline src/ert/storage/local_experiment.py:67]: Nit-picking but I think it's better to do `raise ValueError(f"No {str(self._parameter_file)} exists")`

File: src/ert/_c_wrappers/enkf/config/parameter_config.py [hunk 5]
```diff
@@ -28,3 +41,6 @@ def getKey(self):
 
     def getImplementationType(self):
         return self
+
+    def to_json(self) -> Any:
```
Comment (dafeda) [inline src/ert/_c_wrappers/enkf/config/parameter_config.py:45]: I guess this is to_dict and not to_json 😬

**Review by dafeda — APPROVED** [review]
Looks good 👍

File: src/ert/_c_wrappers/enkf/config/parameter_config.py (hunk 3 above)
Comment (frode-aarstad) [reply src/ert/_c_wrappers/enkf/config/parameter_config.py:14]: The reason for the name is so we can add other custom conversions in the future

Comment (dafeda) [reply src/ert/_c_wrappers/enkf/config/parameter_config.py:14]: A future that may never come, but I see your point.

**Pull request merged**